samcli/lib/samlib/resource_metadata_normalizer.py

```diff
@@ -186,11 +186,9 @@ def _extract_image_asset_metadata(metadata):
         """
         asset_path = Path(metadata.get(ASSET_PATH_METADATA_KEY, ""))
         dockerfile_path = Path(metadata.get(ASSET_DOCKERFILE_PATH_KEY), "")
-        dockerfile, path_from_asset = dockerfile_path.stem, dockerfile_path.parent
-        dockerfile_context = str(Path(asset_path.joinpath(path_from_asset)))
         return {
-            SAM_METADATA_DOCKERFILE_KEY: dockerfile,
-            SAM_METADATA_DOCKER_CONTEXT_KEY: dockerfile_context,
+            SAM_METADATA_DOCKERFILE_KEY: str(dockerfile_path),
```

Comment:
A bit concerned about this change since this will break the current code behaviour, might need to investigate on the current usage volume to decide if we can change this part.

Reply:
> might need to investigate on the current usage volume to decide if we can change this part.

Sounds like a good move 👍 

> this change... will break the current code behaviour

As noted, I think it's good to investigate this.

From our own investigation...

I would not describe this as `breaking` any current behavior.

It does `change` the current behavior. 

But if we say the change `breaks` something... that sounds like: something will stop working.

So - would this `change` cause any currently working code to `break`? 

We do not think it will. 

To illustrate...

Say your CDK code looks like this:

```python
code = aws_lambda.DockerImageCode.from_image_asset(
    directory="/Users/john/PycharmProjects/project-root/application/runtime/",
    file="Dockerfile",
)

docker_lambda = aws_lambda.DockerImageFunction(code=code)
```

This works with the current code.

And this will still work after this PR is merged.

Currently - the SAM CLI only works with CDK code where `file` is just the `Dockerfile` and `directory` is the full path to the `Dockerfile`.

In contrast, CDK code like this...

```python
code = aws_lambda.DockerImageCode.from_image_asset(
    directory="/Users/john/PycharmProjects/project-root",
    file="./application/runtime/Dockerfile",
)

docker_lambda = aws_lambda.DockerImageFunction(code=code)
```

...does not currently work with the SAM CLI. But it will work when this PR is merged.

An important question would be – 

Why was the code written like this to begin with?

In the [PR where this code was introduced](https://github.com/aws/aws-sam-cli/pull/3492) – this does not seem to have been discussed in the comments. 

What this code behaviour specified in your internal tools e.g. Jira? If so, what were the reasons given?

The goal of this feature is to be compatible with the CDK.

But currently, the code is not supporting an important capability of the CDK.

Both `cdk deploy` and `sam build` work fine with metadata like:

```python
DockerContext ="/Users/john/PycharmProjects/project-root",
Dockerfile ="./application/runtime/Dockerfile",
```

So we ware not sure why code has been written to ensure sam always works with metadata like:

```python
DockerContext ="/Users/john/PycharmProjects/project-root/application/runtime",
Dockerfile ="Dockerfile",
```

Reply:
> So it is not clear to us why the code in the PR was been written to ensure sam cli always works with metadata like

This is how SAM CLI handles it. So it was written to match SAM CLI expectations.

Reply:
@jfuss apologies for the slow reply, I was on vacation. 

So I had a look at the SAM CLI code to see how `DockerContext` and `Dockerfile` metadata are used by the SAM CLI.

I found 4 other usages. 

Just one usage is relevant to this PR: 

In [this code block](https://github.com/i18n-tribe/aws-sam-cli/blob/bdb8b961bdcb237105a2776b09cde9944a1f7e5d/samcli/lib/build/app_builder.py#L384-L423) in `samcli/lib/build/app_builder.py` 

We read the metadata properties:

```python
dockerfile = cast(str, metadata.get("Dockerfile"))
docker_context = cast(str, metadata.get("DockerContext"))
```

Then we make `docker_context` into an absolute path:

```python
docker_context_dir = pathlib.Path(self._base_dir, docker_context).resolve()
```

Then we call `docker build` using the python docker client, passing the properties as arguments:

```python
build_args = {
    "path": str(docker_context_dir),
    "dockerfile": dockerfile,
}
build_logs = self._docker_client.api.build(**build_args)
```

So in summary - 

- `Dockerfile` is passed without modification to `_docker_client.api.build` as the `dockerfile` argument
- `DockerContext` is converted to an absolute path, and passed to  `_docker_client.api.build` as the `path` argument

Here is the documentation for ['_docker_client.api.build'](https://docker-py.readthedocs.io/en/stable/api.html#module-docker.api.build)

And those arguments are:

```
path (str) – Path to the directory containing the Dockerfile
dockerfile (str) – path within the build context to the Dockerfile
```

In my opinion, the documentation for `path` is a little bit ambiguous.

Does this mean that `path` is:

1. the immediate parent directory containing the Dockerfile? 
2. a directory that ultimately contains the dockerfile, which might be nested in child folder(s)?

I would completely understand if something thinks it means `1.` - this is what I thought at first.

But if we look at the documentation for `dockerfile` - it says `path within the build context to the Dockerfile`

Because `dockerfile` can be a `path within the build context` - it can be a path (potentially via subfolders) to the `Dockerfile`.

So in fact, `2.` is the correct interpretation.

This interpretation is reinforced by the `docker cli` documentation for [docker build](https://docs.docker.com/engine/reference/commandline/build/)

Usage is:

```
docker build [OPTIONS] PATH 
```

Where `PATH` will be `DockerContext` in the SAM CLI.

And the option `--file` will be `Dockerfile` in the SAM CLI.

In the `docker cli` documentation, they refer to something called the `build context`.

The `build context` is very similar to `PATH` / `DockerContext`.

The `build context` is a copy of the folder specified in `PATH` / `DockerContext`, excluding files specified in `.dockerignore`.

So `--file` / `Dockerfile` is a path to a docker file within the `build context` (usually a relative path is used).

There is no requirement for the docker file to be in the root of the `build context`.

I.E. `--file` / `Dockerfile` can be a path via subfolders to the docker file.

Here's [the documentation for `--file`](https://docs.docker.com/engine/reference/commandline/build/#specify-a-dockerfile--f).

The documentation has examples where `--file` is a path via subfolders to the docker file like:

```
docker build --file dockerfiles/Dockerfile.debug .
docker build --file dockerfiles/Dockerfile.prod .
```

In the example above, `PATH` is `.` - the current directory.


And there's a quote in [this section of the docs](https://docs.docker.com/engine/reference/commandline/build/#text-files).

```
By default the docker build command will look for a Dockerfile at the root of the build context. 
The -f, --file, option lets you specify the path to an alternative file to use instead. 
This is useful in cases where the same set of files are used for multiple builds. 
```

The use case `where the same set of files are used for multiple builds` is exactly what this PR is trying to solve.

This is the use case described in the [PR description](https://github.com/aws/aws-sam-cli/pull/4074#issue-1313551509).

This use-case is supported by the `AWS CDK`.

This use-case is also - in fact - already supported by the `SAM CLI`.

The [`AWS CDK` docs](https://docs.aws.amazon.com/cdk/api/v2/python/aws_cdk.aws_lambda/DockerImageCode.html#aws_cdk.aws_lambda.DockerImageCode.from_image_asset) support this use-case explicitly:

```
directory (str) – the directory from which the asset must be created.
file (Optional[str]) – Path to the Dockerfile (relative to the directory). Default: ‘Dockerfile’
```

The [`SAM CLI` docs](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-using-build.html#build-container-image) do not support this use-case explicitly:

```python
Dockerfile - The name of the Dockerfile associated with the Lambda function.
DockerContext - The location of the Dockerfile.
```

The `SAM CLI` developer guide sound like interpretation `1.` of the documentation for ['_docker_client.api.build'](https://docker-py.readthedocs.io/en/stable/api.html#module-docker.api.build).

But in fact, the `SAM CLI` code works perfectly well if `Dockerfile` is a path via subfolders to the docker file.

And I've personally used this 'undocumented feature' in multiple `SAM` projects I have worked on.

In other words -

This PR is removing code that added a limitation to the `SAM CLI`, when the limitation did not previously exist in either `SAM` or `CDK`, but `SAM CLI` docs said this limitation exists.

So I've created a [PR to update the `SAM CLI` developer guide](https://github.com/awsdocs/aws-sam-developer-guide/pull/159), so people will be aware of this capability of the `SAM CLI`.

I've also created a PR to [update the documentation for the Docker Python Client](https://github.com/docker/docker-py/pull/3043) to be more clear.

---

A further note -

Above there is a link do documentation for the `docker cli`. 

The python docker SDK doesn't use the `docker cli`, it uses the `docker api`.

So I checked the docker API documentation for [build](https://docs.docker.com/engine/api/v1.40/#operation/ImageBuild)

It takes an argument:

```
dockerfile - Path within the build context to the Dockerfile.
```

It also takes a tar archive which contains the `build context` - documented as:

```
Build an image from a tar archive with a Dockerfile in it.

The Dockerfile specifies how the image is built from the tar archive.
It is typically in the archive's root, but can be at a different path or have a different name by specifying the dockerfile parameter.
```

I.e. `dockerfile` 'is typically in the archive's root' - but it can be a path to a dockerfile in a subfolder.

And I checked the source code of the docker python sdk, to see how it's interacting with the docker API.

Here's the [code where it creates the `build context tar archive`](https://github.com/docker/docker-py/blob/b7daa52feb8b897fb10fbe82c6e49a273746352a/docker/api/build.py#L150-L163).

This is a copy of the folder specified in `path` / `DockerContext`, excluding files specified in `.dockerignore`.

And [on this line](https://github.com/docker/docker-py/blob/b7daa52feb8b897fb10fbe82c6e49a273746352a/docker/api/build.py#L174) it passes `Dockerfile` to the docker API as `dockerfile`.
 
---

For reference, here are the 3 other usages of `DockerContext` metadata, which are not relevant to this PR:

- In samcli/commands/build/build_context.py - [If it's a docker lambda, confirm these metadata properties are defined](https://github.com/i18n-tribe/aws-sam-cli/blob/c2b947d6087cd88f2174e6a3b3079865dc06394e/samcli/commands/build/build_context.py#L562-L573).
- In samcli/lib/utils/resource_trigger.py - [trigger a code refresh if this metadata is changed the the developer](https://github.com/i18n-tribe/aws-sam-cli/blob/1cda797850907cf28c7b4f80180cba3627360fd4/samcli/lib/utils/resource_trigger.py#L260-L264)
- In samcli/lib/providers/sam_function_provider.py - [converting `DockerContext` to be an absolute path if necessary](https://github.com/i18n-tribe/aws-sam-cli/blob/4b4e6e44ea7dd7e4d9f73001214dc34a5aa59701/samcli/lib/providers/sam_function_provider.py#L413-L422)

Reply:
Thanks @i18n-tribe for your contribution and your very detailed explanation. I think you have a point regarding this change, I think we miss understood the `DockerContext` and `DockerFile` properties and we assumed that the `DockerFile` should be a file in the `DockerContext` directory path. I did some investigations to make sure that this change will not break any customer. I think it will work as expected, and fix this bug.

my only request is to add some integration testing to cover this case

Reply:
You're most welcome, I'll write some integration tests 👍